torchdp/test/virtual_step_test.py

```diff
@@ -0,0 +1,311 @@
+            loss.backward()
+
+            # accumulate per-sample grads for two mini batches
+            # and then aggregate them in the clipper
+            if (idx + 1) % 2 == 0:
+                self.optimizer.virtual_step()
+
+        # the full data gradient accumulated in .grad
+        grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        ) * (self.BATCH_SIZE / self.DATA_SIZE)
+
+        self.optimizer.step()
+
+        # .grad should contain the average gradient over the entire dataset
+        accumulated_grad = torch.cat(
+            [p.grad.reshape(-1) for p in self.model.parameters() if p.requires_grad]
+        )
+
+        # the accumulated gradients in .grad without any hooks
+        orig_grad = self.effective_batch_grad
+
+        self.assertTrue(
+            torch.allclose(accumulated_grad, orig_grad, atol=10e-5, rtol=10e-3)
+        )
+        self.assertTrue(torch.allclose(grad, orig_grad, atol=10e-5, rtol=10e-3))
+
+    def test_zero_grad(self):
+        """
+        Calling optimizer.zero_grad() should erase any accumulated per-sample gradients.
+        """
+        self.setUp_privacy_engine(2 * self.BATCH_SIZE)
+
+        double_grads_per_sample = []
+        double_grads = []
+
+        for idx, (x, y) in enumerate(self.dl):
+            logits = self.model(x)
+            loss = self.criterion(logits, y)
+            loss.backward()
+
+            # accumulate per-sample grads for two mini batches
+            # and then aggregate them in the clipper
+            if (idx + 1) % 2 == 0:
+
+                # the accumulated per-sample gradients
+                per_sample_grads = torch.cat(
+                    [
+                        p.grad_sample.reshape(2 * self.BATCH_SIZE, -1)
+                        for p in self.model.parameters()
+                        if p.requires_grad
+                    ],
+                    dim=-1,
+                )
+                # average up all the per-sample gradients
+                double_grads_per_sample.append(torch.mean(per_sample_grads, dim=0))
+
+                double_grads.append(
+                    torch.cat(
+                        [
+                            p.grad.reshape(-1)
+                            for p in self.model.parameters()
+                            if p.requires_grad
+                        ]
+                    )
+                    * 0.5
+                )
+
+                self.optimizer.step()
+                self.optimizer.zero_grad()
+
+        double_grads = torch.cat(double_grads)
+        double_grads_per_sample = torch.cat(double_grads_per_sample)
+
+        self.assertTrue(
+            torch.allclose(
+                self.double_grads, double_grads_per_sample, atol=10e-5, rtol=10e-3
+            )
+        )
+        self.assertTrue(
+            torch.allclose(self.double_grads, double_grads, atol=10e-5, rtol=10e-3)
```

Comment:
Hmm I am now a bit confused by this. :) We are checking if `self.double_grads` is close to `double_grads`. Should we use a better naming for these variables? (comparing `x` and `self.x` may not be very clear)

Reply:
as above